src/tangents.jl

```diff
@@ -55,6 +55,13 @@ Base.:(==)(x::Tangent, y::Tangent) = x.fields == y.fields
 
 mutable struct MutableTangent{Tfields<:NamedTuple}
     fields::Tfields
+    MutableTangent{Tfields}(fields::Tfields) where {Tfields} = new{Tfields}(fields)
+    function MutableTangent{Tfields}(fields::NamedTuple{names}) where {Tfields, names}
+        @assert names == fieldnames(Tfields) "field names mismatch"
+        new{Tfields}(fields)
+    end
```

Comment:
Would you consider using diagonal dispatch here, rather than error checking? Something like
```julia
    MutableTangent{Tfields}(fields::Tfields) where {Tfields} = new{Tfields}(fields)
```

Reply:
Ahh right, yes, I see why this is needed -- I'm very glad we have a decent number of test cases now!

Perhaps this would work:
```suggestion
    function MutableTangent{Tfields}(fields::NamedTuple{names}) where {names, Tfields<:NamedTuple{names}}
        return new{Tfields}(fields)
    end
```